File: Grocery_Sense/GrocerySense.Data/Repositories/ItemsAdminRepo.cs
using System.Collections.Immutable;
using GrocerySense.Domain;
using Microsoft.Data.Sqlite;

namespace GrocerySense.Data.Repositories;

// Port of reference-python/.../data/repositories/items_admin_repo.py — the Item Manager admin ops:
// search, rename, and safe merge (re-point every item_id-bearing table). Adds CorrectLineMapping, the
// receipt-line alias-correction (fix line + learn) that has no Python source.
//
// No ensure_schema: items.is_tracked/default_unit come from the migration ledger, not a runtime ALTER.
// Merge + correction MUST be atomic (they touch many tables), so those take a required transaction; the
// caller owns it (matches ReceiptsRepo.IngestReceipt).
public static class ItemsAdminRepo
{
    // Canonical unit vocabulary (lowercased), mirroring UnitNormalizationService.NormalizeUnit's outputs.
    // Duplicated here because Data cannot reference Core; a merged item's default_unit is only ever set from
    // that vocabulary, so a too-narrow whitelist would silently drop a valid unit like "l" during promotion.
    private static readonly string[] ValidUnits =
        { "each", "lb", "kg", "g", "oz", "l", "ml", "fl_oz", "cup", "tbsp", "tsp", "gal", "pint", "dozen", "bunch", "case", "pack" };

    // Every table with an item_id column referencing items.id, re-pointed on merge. Verified against the
    // live schema (Database.cs), NOT the Python list — flyer_deals/price_drop_alerts/watchlist post-date it.
    // No table here has a UNIQUE on item_id, so a plain UPDATE never collides. watchlist is handled
    // separately below: it has no UNIQUE(item_id) either, so a blind UPDATE would leave the target with two
    // active watches. Any new item_id table MUST be added here (or to the watchlist special-case).
    // internal (not private) so a Tests schema-drift guard can assert this list plus watchlist covers
    // EVERY item_id-bearing table in the live schema — a new one added to neither orphans its rows on merge.
    // ImmutableArray (not string[]): now that it's exposed beyond this class, readonly alone guards only the
    // reference — array elements would still be reassignable. The names feed interpolated UPDATE SQL, so pin them.
    internal static readonly ImmutableArray<string> ItemIdTables =
        ["prices", "receipt_line_items", "shopping_list", "flyer_deals", "price_drop_alerts", "item_aliases"];

    private static readonly ItemAliasesRepo Aliases = new();

    // Items with light price stats, newest-tracked first. Optional case-insensitive name filter.
    public static IReadOnlyList<ItemAdminRow> SearchItems(SqliteConnection conn, string query = "",
        int limit = 250, SqliteTransaction? tx = null)
    {
        var q = (query ?? "").Trim();
        // Escape LIKE metacharacters so a query like "2% milk" doesn't become a wildcard that over-matches the
        // merge picker (picking the wrong item destructively merges price history).
        var where = q.Length > 0 ? @"WHERE i.canonical_name LIKE $q ESCAPE '\'" : "";
        // Pick the (bounded) matching items FIRST, then aggregate prices only for those ids. The old shape
        // GROUP BY'd the entire prices table before applying the LIMIT, so search cost grew with total price
        // history rather than with the handful of rows actually returned.
        using var cmd = Db.Command(conn, tx, $"""
            WITH selected AS (
                SELECT i.id, i.canonical_name, i.is_tracked, i.default_unit
                FROM items i
                {where}
                ORDER BY COALESCE(i.is_tracked, 0) DESC, i.canonical_name ASC
                LIMIT $limit
            ),
            price_stats AS (
                SELECT p.item_id, COUNT(1) AS price_points, MAX(p.date) AS last_price_date
                FROM prices p JOIN selected s ON s.id = p.item_id
                GROUP BY p.item_id
            )
            SELECT s.id, COALESCE(s.canonical_name, ''), COALESCE(s.is_tracked, 0), s.default_unit,
                   COALESCE(ps.price_points, 0), ps.last_price_date
            FROM selected s
            LEFT JOIN price_stats ps ON ps.item_id = s.id
            ORDER BY COALESCE(s.is_tracked, 0) DESC, s.canonical_name ASC
            """);
        if (q.Length > 0) cmd.Parameters.AddWithValue("$q", $"%{EscapeLike(q)}%");
        cmd.Parameters.AddWithValue("$limit", limit);

        using var r = cmd.ExecuteReader();
        var rows = new List<ItemAdminRow>();
        while (r.Read())
            rows.Add(new ItemAdminRow(r.GetInt32(0), r.GetString(1), r.GetInt32(2) != 0,
                r.GetStringOrNull(3)?.Trim().ToLowerInvariant(), r.GetInt32(4), r.GetStringOrNull(5)));
        return rows;
    }

    // Rename an item's canonical name. canonical_name is UNIQUE, so colliding with an existing name is a
    // constraint error -> surface it as a clear message rather than a raw SqliteException.
    public static void RenameItem(SqliteConnection conn, int itemId, string newName, SqliteTransaction? tx = null)
    {
        var name = (newName ?? "").Trim();
        if (name.Length == 0) throw new ArgumentException("New name cannot be empty.", nameof(newName));

        using var cmd = Db.Command(conn, tx, "UPDATE items SET canonical_name = $name WHERE id = $id");
        cmd.Parameters.AddWithValue("$name", name);
        cmd.Parameters.AddWithValue("$id", itemId);
        try { cmd.ExecuteNonQuery(); }
        catch (SqliteException e) when (e.SqliteErrorCode == 19)
        {
            throw new InvalidOperationException($"An item named \"{name}\" already exists — merge into it instead.");
        }
    }

    // Merge source -> target: move every item_id reference, promote tracked/default_unit, keep the source
    // name as an alias, delete source. All within the caller's transaction (all-or-nothing).
    public static void MergeItems(SqliteConnection conn, SqliteTransaction tx, int targetItemId, int sourceItemId,
        bool keepSourceAsAlias = true)
    {
        if (targetItemId == sourceItemId) throw new ArgumentException("Target and source item are the same.");

        var target = ItemsRepo.GetItemById(conn, targetItemId, tx)
            ?? throw new InvalidOperationException($"Target item not found: {targetItemId}");
        var source = ItemsRepo.GetItemById(conn, sourceItemId, tx)
            ?? throw new InvalidOperationException($"Source item not found: {sourceItemId}");

        // Promote: if either is tracked, target is tracked; adopt source's default unit if target has none.
        if (!target.IsTracked && source.IsTracked)
            Exec(conn, tx, "UPDATE items SET is_tracked = 1 WHERE id = $id", ("$id", targetItemId));

        var sourceUnit = source.DefaultUnit?.Trim().ToLowerInvariant();
        if (string.IsNullOrEmpty(target.DefaultUnit) && sourceUnit is not null && ValidUnits.Contains(sourceUnit))
            Exec(conn, tx, "UPDATE items SET default_unit = $u WHERE id = $id",
                ("$u", sourceUnit), ("$id", targetItemId));

        // Move references. No item_id UNIQUE on these tables, so a straight UPDATE is safe.
        foreach (var table in ItemIdTables)
            Exec(conn, tx, $"UPDATE {table} SET item_id = $t WHERE item_id = $s",
                ("$t", targetItemId), ("$s", sourceItemId));

        // watchlist: prevent two ACTIVE watches for the merged item. Only drop the source's watch when the
        // target already has an active one — scoping on is_active=1 so a paused (soft-deleted) target watch
        // never causes the source's live watch to be silently deleted.
        Exec(conn, tx,
            "DELETE FROM watchlist WHERE item_id = $s AND EXISTS (SELECT 1 FROM watchlist WHERE item_id = $t AND is_active = 1)",
            ("$s", sourceItemId), ("$t", targetItemId));
        Exec(conn, tx, "UPDATE watchlist SET item_id = $t WHERE item_id = $s",
            ("$t", targetItemId), ("$s", sourceItemId));

        // Keep the source name discoverable: its canonical name becomes an alias of the target.
        if (keepSourceAsAlias && !string.IsNullOrWhiteSpace(source.CanonicalName))
            Aliases.UpsertAlias(conn, source.CanonicalName, targetItemId, 1.0, "merge", tx);

        Exec(conn, tx, "DELETE FROM items WHERE id = $id", ("$id", sourceItemId));
    }

    // Alias-correction ("fix line + learn"): re-point one receipt line and the price row it produced to the
    // correct item, and learn description -> item so future scans map right. No retro-sweep — earlier
    // mis-mapped receipts are cleaned via MergeItems. All in the caller's transaction.
    // ponytail: an originally-unmapped line (oldItemId null) has no price row to move; it still re-points the
    // line + learns the alias. Back-creating the missing historical price is out of scope (re-import covers it).
    public static void CorrectLineMapping(SqliteConnection conn, SqliteTransaction tx, int lineItemId,
        int receiptId, string description, int? oldItemId, int newItemId)
    {
        Exec(conn, tx, "UPDATE receipt_line_items SET item_id = $new WHERE id = $line",
            ("$new", newItemId), ("$line", lineItemId));

        // The price row from this line is keyed (receipt_id, raw_name = description, item_id = old). Move ONLY
        // ONE matching row: two identical-description lines on the same receipt map to identical price rows, so
        // an unbounded UPDATE would move both when fixing the first line, orphaning the second correction.
        // ponytail: no line_item_id link on prices; single-row move keeps counts/attribution correct without a
        // schema migration (identical rows make the exact pairing immaterial). Add line_item_id if that changes.
        if (oldItemId is int old)
            Exec(conn, tx,
                """
                UPDATE prices SET item_id = $new WHERE id = (
                    SELECT id FROM prices
                    WHERE receipt_id = $rid AND raw_name = $raw AND item_id = $old
                    ORDER BY id LIMIT 1)
                """,
                ("$new", newItemId), ("$rid", receiptId), ("$raw", description), ("$old", old));

        Aliases.UpsertAlias(conn, description, newItemId, 1.0, "manual_correction", tx);
    }

    // Escape LIKE wildcards for use with ESCAPE '\'. Backslash first so we don't double-escape our own escapes.
    private static string EscapeLike(string s) =>
        s.Replace("\\", "\\\\").Replace("%", "\\%").Replace("_", "\\_");

    private static void Exec(SqliteConnection conn, SqliteTransaction? tx, string sql,
        params (string Name, object Value)[] ps)
    {
        using var cmd = Db.Command(conn, tx, sql);
        foreach (var (name, value) in ps) cmd.Parameters.AddWithValue(name, value);
        cmd.ExecuteNonQuery();
    }
}

File: reference-python/src/Grocery_Sense/data/repositories/items_admin_repo.py
from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

from Grocery_Sense.data.connection import connection_scope, get_connection


def _now_utc_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


VALID_UNITS = ("each", "lb", "kg", "g")

# Tables this admin repo is allowed to inspect via PRAGMA. PRAGMA cannot accept
# `?` placeholders, so all dynamic table names go through this whitelist.
_ALLOWED_PRAGMA_TABLES = frozenset({
    "items", "item_aliases", "prices", "receipt_line_items",
    "shopping_list", "flyer_deals",
})

# Tables that carry an item_id column referencing items.id. Used by merge_items
# instead of walking sqlite_master (per CLAUDE.md: no generic helpers for
# one-off operations; new tables must be added here explicitly).
_ITEM_ID_TABLES = ("prices", "receipt_line_items", "shopping_list", "flyer_deals", "item_aliases")


@dataclass(frozen=True)
class ItemRow:
    id: int
    canonical_name: str
    is_tracked: int
    default_unit: Optional[str]
    price_points: int
    last_price_date: Optional[str]


class ItemsAdminRepo:
    """
    Admin helpers for Item Manager UI.

    - Ensures items.default_unit + items.is_tracked exist
    - Searches items
    - Toggles tracked
    - Sets default unit
    - Renames canonical name
    - Merges items safely by updating all tables containing item_id
    """

    # ---------------------------
    # Schema ensure
    # ---------------------------

    def ensure_schema(self) -> None:
        self._ensure_items_columns()

    def _col_exists(self, table: str, col: str) -> bool:
        if table not in _ALLOWED_PRAGMA_TABLES:
            raise ValueError(f"Disallowed table for PRAGMA: {table}")
        with connection_scope() as conn:
            rows = conn.execute(f"PRAGMA table_info({table});").fetchall()
        return any(r[1] == col for r in rows)

    def _ensure_items_columns(self) -> None:
        with connection_scope() as conn:
            # items.is_tracked
            if not self._col_exists("items", "is_tracked"):
                conn.execute("ALTER TABLE items ADD COLUMN is_tracked INTEGER NOT NULL DEFAULT 0;")

            # items.default_unit
            if not self._col_exists("items", "default_unit"):
                conn.execute("ALTER TABLE items ADD COLUMN default_unit TEXT;")

            conn.commit()

    # ---------------------------
    # Queries
    # ---------------------------

    def search_items(self, query: str = "", *, limit: int = 250) -> List[ItemRow]:
        """
        Returns items with light stats (price_points + last_price_date).
        """
        self.ensure_schema()
        q = (query or "").strip()

        where = ""
        params: List[Any] = []
        if q:
            where = "WHERE i.canonical_name LIKE ?"
            params.append(f"%{q}%")

        sql = f"""
            SELECT
                i.id,
                COALESCE(i.canonical_name, '') AS canonical_name,
                COALESCE(i.is_tracked, 0) AS is_tracked,
                i.default_unit,
                COALESCE(ps.price_points, 0) AS price_points,
                ps.last_price_date
            FROM items i
            LEFT JOIN (
                SELECT item_id, COUNT(1) AS price_points, MAX(date) AS last_price_date
                FROM prices
                GROUP BY item_id
            ) ps ON ps.item_id = i.id
            {where}
            ORDER BY COALESCE(i.is_tracked, 0) DESC, i.canonical_name ASC
            LIMIT ?;
        """
        params.append(int(limit))

        with connection_scope() as conn:
            rows = conn.execute(sql, tuple(params)).fetchall()

        out: List[ItemRow] = []
        for r in rows:
            out.append(
                ItemRow(
                    id=int(r[0]),
                    canonical_name=str(r[1] or ""),
                    is_tracked=int(r[2] or 0),
                    default_unit=(str(r[3]).strip().lower() if r[3] else None),
                    price_points=int(r[4] or 0),
                    last_price_date=r[5],
                )
            )
        return out

    def get_item(self, item_id: int) -> Optional[Dict[str, Any]]:
        self.ensure_schema()
        with connection_scope() as conn:
            r = conn.execute(
                """
                SELECT id, canonical_name, COALESCE(is_tracked,0), default_unit
                FROM items
                WHERE id = ?;
                """,
                (int(item_id),),
            ).fetchone()
        if not r:
            return None
        return {
            "id": int(r[0]),
            "canonical_name": r[1],
            "is_tracked": int(r[2] or 0),
            "default_unit": (str(r[3]).strip().lower() if r[3] else None),
        }

    # ---------------------------
    # Mutations
    # ---------------------------

    def toggle_tracked(self, item_id: int) -> int:
        self.ensure_schema()
        with connection_scope() as conn:
            cur = conn.execute("SELECT COALESCE(is_tracked,0) FROM items WHERE id=?;", (int(item_id),)).fetchone()
            if not cur:
                raise ValueError(f"Item not found: {item_id}")
            new_val = 0 if int(cur[0] or 0) == 1 else 1
            conn.execute("UPDATE items SET is_tracked=? WHERE id=?;", (new_val, int(item_id)))
            conn.commit()
        return new_val

    def set_default_unit(self, item_id: int, default_unit: Optional[str]) -> None:
        self.ensure_schema()
        unit = (default_unit or "").strip().lower() or None
        if unit is not None and unit not in VALID_UNITS:
            raise ValueError(f"Invalid unit: {unit} (allowed: {', '.join(VALID_UNITS)})")

        with connection_scope() as conn:
            conn.execute("UPDATE items SET default_unit=? WHERE id=?;", (unit, int(item_id)))
            conn.commit()

    def rename_item(self, item_id: int, new_name: str) -> None:
        self.ensure_schema()
        name = (new_name or "").strip()
        if not name:
            raise ValueError("New name cannot be empty.")

        with connection_scope() as conn:
            conn.execute("UPDATE items SET canonical_name=? WHERE id=?;", (name, int(item_id)))
            conn.commit()

    def merge_items(
        self,
        *,
        target_item_id: int,
        source_item_id: int,
        keep_source_as_alias: bool = True,
    ) -> None:
        """
        Merge source -> target:
          - move all references (tables with item_id column) from source to target
          - copy default_unit to target if target missing
          - OR tracked: if either tracked, target becomes tracked
          - delete source row

        If keep_source_as_alias, attempts to insert source canonical name into item_aliases
        (only if the table exists).
        """
        self.ensure_schema()

        if int(target_item_id) == int(source_item_id):
            raise ValueError("Target and source item are the same.")

        with connection_scope() as conn:
            conn.execute("BEGIN;")
            try:
                t = conn.execute(
                    "SELECT id, canonical_name, COALESCE(is_tracked,0), default_unit FROM items WHERE id=?;",
                    (int(target_item_id),),
                ).fetchone()
                s = conn.execute(
                    "SELECT id, canonical_name, COALESCE(is_tracked,0), default_unit FROM items WHERE id=?;",
                    (int(source_item_id),),
                ).fetchone()

                if not t:
                    raise ValueError(f"Target item not found: {target_item_id}")
                if not s:
                    raise ValueError(f"Source item not found: {source_item_id}")

                target_tracked = int(t[2] or 0)
                source_tracked = int(s[2] or 0)
                target_unit = (str(t[3]).strip().lower() if t[3] else None)
                source_unit = (str(s[3]).strip().lower() if s[3] else None)

                # Promote tracked / default_unit
                if target_tracked == 0 and source_tracked == 1:
                    conn.execute("UPDATE items SET is_tracked=1 WHERE id=?;", (int(target_item_id),))

                if (not target_unit) and source_unit in VALID_UNITS:
                    conn.execute("UPDATE items SET default_unit=? WHERE id=?;", (source_unit, int(target_item_id)))

                # Move references across the known item_id-bearing tables. New
                # such tables must be added to _ITEM_ID_TABLES explicitly.
                existing_tables = {
                    r[0] for r in conn.execute(
                        "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%';"
                    ).fetchall()
                }

                for table in _ITEM_ID_TABLES:
                    if table not in existing_tables:
                        continue

                    # Wrap per-table UPDATE in a SAVEPOINT so a unique-constraint
                    # collision only rolls back this table's change, leaving the
                    # outer transaction intact.
                    conn.execute("SAVEPOINT merge_table;")
                    try:
                        conn.execute(
                            f"UPDATE {table} SET item_id=? WHERE item_id=?;",
                            (int(target_item_id), int(source_item_id)),
                        )
                        conn.execute("RELEASE merge_table;")
                    except sqlite3.IntegrityError:
                        conn.execute("ROLLBACK TO merge_table;")
                        conn.execute("RELEASE merge_table;")
                        conn.execute(
                            f"DELETE FROM {table} WHERE item_id=?;",
                            (int(source_item_id),),
                        )

                # Keep source name as alias (optional, if item_aliases exists)
                if keep_source_as_alias:
                    source_name = (s[1] or "").strip().lower()
                    if source_name:
                        if "item_aliases" in existing_tables:
                            # best-effort schema: alias_text, item_id, confidence, source, created_at
                            cols = conn.execute("PRAGMA table_info(item_aliases);").fetchall()
                            alias_cols = {c[1] for c in cols}
                            if {"alias_text", "item_id"}.issubset(alias_cols):
                                # avoid duplicates
                                existing = conn.execute(
                                    "SELECT 1 FROM item_aliases WHERE alias_text=? LIMIT 1;",
                                    (source_name,),
                                ).fetchone()
                                if not existing:
                                    # flexible insert based on available cols
                                    fields = ["alias_text", "item_id"]
                                    values = [source_name, int(target_item_id)]

                                    if "confidence" in alias_cols:
                                        fields.append("confidence")
                                        values.append(1.0)
                                    if "source" in alias_cols:
                                        fields.append("source")
                                        values.append("merge")
                                    if "created_at" in alias_cols:
                                        fields.append("created_at")
                                        values.append(_now_utc_iso())

                                    placeholders = ", ".join(["?"] * len(values))
                                    conn.execute(
                                        f"INSERT INTO item_aliases ({', '.join(fields)}) VALUES ({placeholders});",
                                        tuple(values),
                                    )

                # Delete the source item
                conn.execute("DELETE FROM items WHERE id=?;", (int(source_item_id),))

                conn.commit()

            except Exception:
                conn.execute("ROLLBACK;")
                raise
